site/content/entry/yuzu-fastmem/index.md

```diff
@@ -0,0 +1,118 @@
++++
+date = "2021-06-08T03:00:00+05:30"
+title = "New Feature Release - Fastmem Support"
+author = "CaptV0rt3x"
+forum = 415882
++++
+
+Hey there, yuz-ers!
+While all of you wait eagerly for the release of Project Hades, our shader decompiler rewrite, we thought we'd bring you a nice little surprise to keep you occupied. 
+We present to you, the newest addition to yuzu's ever-improving features list - Fastmem (Fast Memory Access)!
+
+<!--more-->
+&nbsp;
+
+This is now available in the latest yuzu Early Access build (1759) and we intend to make this available in mainline builds very soon. 
+As always, we ask that you test various games with these builds and if you encounter any issues, bugs, or crashes, please reach out to us via the [Discord](https://discord.gg/u77vRWY) Patreon channels.
+
+**Note: This feature is enabled by default.**
+
+## What is Fastmem?
+
+Fastmem or Fast Memory Access, is a well-known feature among emulator developers and emulation enthusiasts. 
+Emulators like Dolphin and Citra Android have implemented this in the past.
+
+All modern Operating Systems have this concept of "virtual memory," where addressable memory (from the running process) is a space of memory addresses that are only "known" to that process, e.g. Discord on your PC, or Super Mario Odyssey on your Switch — and the addresses are translated by the Switch's MMU (Memory Management Unit) to the physical location in the Switch’s physical RAM as the game runs.
+
+&nbsp;
+
+{{< imgs
+	"./MMU.png|Illustration of an MMU (By Mdjango, Andrew S. Tanenbaum)"
+>}}
+
+In yuzu, we emulate the Switch's virtual address space for the running process. 
+But, that means for each time the game tries to read/write memory, which happens millions of times per frame, we need to basically "decode" where that Switch virtual address maps to our "emulated" Switch memory (RAM allocated for yuzu).
+This adds a lot of overhead when you consider how much the game reads/writes to memory.
+
+The idea behind fastmem here is to map the emulated Switch game's addressable memory on the host (Windows / Linux), at offset addresses. 
+The offset is a constant.
+
+## The Technical - HOW?
+
+On the Switch, we have 4GB of memory. 
+The hardware supports a virtual memory size of 48bits (just like x64); however, the Switch kernel limits this to only 39bit to save some memory for storing page tables.
+This results in a 512GB virtual address space (2^39) for use. (A page table is the data structure used to store the mapping between virtual addresses and physical addresses.)
+
+To emulate this, we allocated a HUGE page table, which had an offset for each memory page addressable in the virtual memory region of the Switch. Just this table was 1GB! 
```

Comment:
```suggestion
To emulate this, we allocated a HUGE page table, which had an offset for each memory page addressable in the virtual memory region of the Switch. This table alone was 1GB in size! 
```
Didn't like how I phrased this before lol

Reply:
ah ha!